staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -674,7 +675,221 @@ type crdConversionRESTOptionsGetter struct {
+	obj, gvk, err := d.delegate.Decode(data, defaults, into)
+	if err != nil {
+		return nil, gvk, err
+	}
+	if u, ok := obj.(*unstructured.Unstructured); ok {
+		if err := d.validator.apply(u); err != nil {
+			return nil, gvk, err
+		}
+	}
+
+	return obj, gvk, nil
+}
+
+// validatingConverter calls the delegate converter and applies the Unstructured validator.
+type validatingConverter struct {
+	delegate  runtime.ObjectConvertor
+	validator unstructuredSchemaValidator
+}
+
+var _ runtime.ObjectConvertor = validatingConverter{}
+
+func (v validatingConverter) Convert(in, out, context interface{}) error {
+	if err := v.delegate.Convert(in, out, context); err != nil {
+		return err
+	}
+
+	if u, ok := out.(*unstructured.Unstructured); ok {
+		if err := v.validator.apply(u); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (v validatingConverter) ConvertToVersion(in runtime.Object, gv runtime.GroupVersioner) (runtime.Object, error) {
+	out, err := v.delegate.ConvertToVersion(in, gv)
+	if err != nil {
+		return nil, err
+	}
+
+	if u, ok := out.(*unstructured.Unstructured); ok {
+		if err := v.validator.apply(u); err != nil {
+			return nil, err
+		}
+	}
+
+	return out, nil
+}
+
+func (v validatingConverter) ConvertFieldLabel(version, kind, label, value string) (string, string, error) {
+	return v.ConvertFieldLabel(version, kind, label, value)
+}
+
+// unstructuredSchemaValidator does the validation for Unstructured that json.Unmarshal
+// does for native types. This includes:
+// - validating and pruning ObjectMeta (here with optional error instead of pruning)
+// - TODO: application of an OpenAPI validator (against the whole object or a top-level field of it).
+// - TODO: optionally application of post-validation algorithms like defaulting and/or OpenAPI based pruning.
+type unstructuredSchemaValidator struct {
+	dropInvalidMetadata bool
+}
+
+func (v *unstructuredSchemaValidator) apply(u *unstructured.Unstructured) error {
+	// save implicit meta fields that don't have to be specified in the validation spec
+	kind, _, err := unstructured.NestedString(u.UnstructuredContent(), "kind")
+	if err != nil {
+		return err
+	}
+	apiVersion, _, err := unstructured.NestedString(u.UnstructuredContent(), "apiVersion")
+	if err != nil {
+		return err
+	}
+
+	if v.dropInvalidMetadata {
+		dropInvalidTypedFields(u)
+	}
+
+	objectMeta, err := getObjectMeta(u)
+	if err != nil {
+		return err
+	}
+
+	// restore meta fields, starting clean
+	u.UnstructuredContent()["metadata"] = map[string]interface{}{}
+	u.SetKind(kind)
+	u.SetAPIVersion(apiVersion)
+	setObjectMeta(u, objectMeta)
+
+	return nil
+}
+
+// dropInvalidTypedFields removes fields with wrong types, avoiding them creating more harm.
+// Only list those fields where it is nore important to keep the objects available to the user
+// than preserving all (invalid) fields.
+func dropInvalidTypedFields(u *unstructured.Unstructured) {
```

Comment:
@liggitt please check that this behaviour is sane. Those fields left out here will fail in the derialization into `ObjectMeta` further down.

I keep using JSON deserialization for transformation into `ObjectMeta`, because it gives us the schema validation. The object accessors for `unstructured.Unstructured` are too forgiving, and especially in unexpected ways ("best effort").

Reply:
why wouldn't we handle all fields here (including name, namespace, resourceVersion, creationTimestamp, deletionTimestamp)? I expected dropping invalid fields to *always* result in a convertible map, even at the expense of missing required fields (which validation would catch)

Reply:
also, I'm unsure about the fuzzing test... it doesn't seem to be reliably populating generation, initializers, ownerReferences, namespace, deletion timestamp, etc

and when I populate those things, we can still get failures after supposedly dropping invalid fields

see https://gist.github.com/liggitt/b6ba8632edf9ea9c1ed54c3a7ae7a7f5

```
--- FAIL: TestRoundtripObjectMeta (0.02s)
	customresource_handler_test.go:186: diff: 
		object.CreationTimestamp.Time:
		  a: ...e.Time{wall:0xbeba287aa121afca, ext:35706598, loc:(*time.Location)(0x2a72400)}"
		  b: ...e.Time{wall:0x0, ext:63662858602, loc:(*time.Location)(0x2a72400)}"
		object.DeletionTimestamp.Time:
		  a: ...e.Time{wall:0xbeba287aa121b1d0, ext:35707116, loc:(*time.Location)(0x2a72400)}"
		  b: ...e.Time{wall:0x0, ext:63662858602, loc:(*time.Location)(0x2a72400)}"
		Codec: &json.Serializer{meta:json.SimpleMetaFactory{}, creater:(*runtime.Scheme)(0xc4200e7ce0), typer:(*runtime.Scheme)(0xc4200e7ce0), yaml:false, pretty:false}
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences=1: invalid metadata: cannot restore slice from int64
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences=1.5: invalid metadata: cannot restore slice from float64
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences=true: invalid metadata: cannot restore slice from bool
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences="a": invalid metadata: cannot restore slice from string
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences=[]string{"a", "b"}: invalid metadata: cannot restore struct from: string
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences=map[string]string{"a":"1", "b":"2"}: invalid metadata: cannot restore slice from map
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences=[]int64{1, 2}: invalid metadata: cannot restore struct from: int64
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences=[]float64{1.5, 2.5}: invalid metadata: cannot restore struct from: float64
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences=map[string]interface {}{"b":interface {}(nil), "a":"1"}: invalid metadata: cannot restore slice from map
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences=map[string]interface {}{"a":"1", "b":[]string{"nested"}}: invalid metadata: cannot restore slice from map
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on ownerReferences=[]interface {}{"a", 1, 1.5, true, []string{"nested"}}: invalid metadata: cannot restore struct from: string
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on initializers=1: invalid metadata: cannot restore struct from: int64
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on initializers=1.5: invalid metadata: cannot restore struct from: float64
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on initializers=true: invalid metadata: cannot restore struct from: bool
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on initializers="a": invalid metadata: cannot restore struct from: string
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on initializers=[]string{"a", "b"}: invalid metadata: cannot restore struct from: slice
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on initializers=[]int64{1, 2}: invalid metadata: cannot restore struct from: slice
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on initializers=[]float64{1.5, 2.5}: invalid metadata: cannot restore struct from: slice
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on initializers=[]interface {}{"a", 1, 1.5, true, []string{"nested"}}: invalid metadata: cannot restore struct from: slice
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on deletionTimestamp="a": invalid metadata: parsing time "a" as "2006-01-02T15:04:05Z07:00": cannot parse "a" as "2006"
	customresource_handler_test.go:254: labels=map[string]string{"a":"1", "b":"2"}, diff: 
		object.Labels[a]:
		  a: "1"
		  b: <nil>
		object.Labels[b]:
		  a: "2"
		  b: <nil>
	customresource_handler_test.go:254: labels=map[string]interface {}{"a":"1"}, diff: 
		object.Labels[b]:
		  a: ""
		  b: <nil>
	customresource_handler_test.go:254: labels=map[string]interface {}{"a":"1"}, diff: 
		object.Labels[a]:
		  a: <nil>
		  b: "1"
	customresource_handler_test.go:249: got unexpected error after dropping invalid typed fields on creationTimestamp="a": invalid metadata: parsing time "a" as "2006-01-02T15:04:05Z07:00": cannot parse "a" as "2006"
	customresource_handler_test.go:254: annotations=map[string]string{"a":"1", "b":"2"}, diff: 
		object.Annotations[a]:
		  a: "1"
		  b: <nil>
		object.Annotations[b]:
		  a: "2"
		  b: <nil>
	customresource_handler_test.go:254: annotations=map[string]interface {}{"a":"1"}, diff: 
		object.Annotations[b]:
		  a: ""
		  b: <nil>
	customresource_handler_test.go:254: annotations=map[string]interface {}{"a":"1"}, diff: 
		object.Annotations[a]:
		  a: <nil>
		  b: "1"
	customresource_handler_test.go:254: finalizers=[]string{"a", "b"}, diff: 
		object.Finalizers[0]:
		  a: "a"
		  b: <nil>
		object.Finalizers[1]:
		  a: "b"
		  b: <nil>
	customresource_handler_test.go:254: finalizers=[]interface {}{"a", 1, 1.5, true, []string{"nested"}}, diff: 
		object.Finalizers[0]:
		  a: <nil>
		  b: "a"
FAIL
FAIL	k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver	0.100s
```

Reply:
So our fuzzers are incomplete :-/

Reply:
> why wouldn't we handle all fields here

It's a choice between consistent objects (we don't validate on storage reads, do we?) and availability of existing objects. An object without name is really really really broken. `initializers` are alpha (didn't we want to remove them in 1.11?).

I can add the missing fields. Not really sure which side of the consistency-availablity scale we want.

Reply:
I think the result of dropping invalid fields should be an object that can *always* successfully convert to ObjectMeta. Dropping field by field like https://gist.github.com/liggitt/62f5674f11d0e8d0638fe2d5baec5361 would let us preserve all well-structured fields.